xcodeproj/internal/xcschemes/xcscheme_infos.bzl

```diff
@@ -0,0 +1,684 @@
+is, make sure it's a transitive dependency of a top-level target in the \
+`xcodeproj.top_level_targets` attribute.
+""".format(label = label, scheme = scheme_name),
+        )
+
+    return target_id
+
+def _get_target_ids(
+        *,
+        scheme_name,
+        target_environment,
+        top_level_deps,
+        top_level_label,
+        xcode_configuration):
+    if not target_environment:
+        if "simulator" in top_level_deps:
+            target_environment = "simulator"
+        else:
+            target_environment = "device"
+    target_ids_by_configuration = top_level_deps.get(target_environment)
+    if not target_ids_by_configuration:
+        fail(
+            """\
+Unknown target environment in `xcscheme` "{scheme}": {env}
+""".format(env = target_environment, scheme = scheme_name),
+        )
+
+    target_ids_by_label = target_ids_by_configuration.get(xcode_configuration)
+    if not target_ids_by_label:
+        fail(
+            """\
+Unknown Xcode configuration in `xcscheme` "{scheme}": {config}
+""".format(config = xcode_configuration, scheme = scheme_name),
+        )
+
+    target_ids = target_ids_by_label.get(top_level_label)
+    if not target_ids:
+        fail(
+            """\
+Unknown top-level target in `xcscheme` "{scheme}": {label}
+
+Is '{label}' an `alias` target? Only actual target labels are support in \
+`xcscheme` definitions. Check that '{label}' is spelled correctly, and if it \
+is, make sure it's in the `xcodeproj.top_level_targets` attribute.
+""".format(label = top_level_label, scheme = scheme_name),
+        )
+
+    return target_ids
+
+def _get_top_level_id(
+        *,
+        scheme_name,
+        target_environment,
+        top_level_deps,
+        top_level_label,
+        xcode_configuration):
+    target_ids = _get_target_ids(
+        scheme_name = scheme_name,
+        target_environment = target_environment,
+        top_level_deps = top_level_deps,
+        top_level_label = top_level_label,
+        xcode_configuration = xcode_configuration,
+    )
+    return target_ids.id
+
+def _launch_target_info_from_dict(
+        launch_target,
+        *,
+        scheme_name,
+        top_level_deps,
+        xcode_configuration):
+    if not launch_target:
+        return (
+            _make_launch_target(),
+            [],
+        )
+
+    if type(launch_target) == "string":
+        return (
+            _make_launch_target(
+                extension_host = EMPTY_STRING,
+                id = _get_top_level_id(
+                    target_environment = None,
+                    top_level_deps = top_level_deps,
+                    top_level_label = launch_target,
+                    xcode_configuration = xcode_configuration,
+                ),
+                post_actions = [],
+                pre_actions = [],
+                working_directory = EMPTY_STRING,
+            ),
+            [],
+        )
+
+    target_ids = _get_target_ids(
```

Comment:
Q: This is plural because it's one ID per configuration correct? If that's the case when you say `target_ids.id` below how is `.id` picked?

Reply:
The function and variable were poorly named. I've renamed them.